### ISSUE TITLE
Kubernetes pods not respecting /etc/resolv.conf nameserver order

### ISSUE THREAD

**Issue by chrisghill** [kubernetes/kubernetes#78138]
**What happened**:
TL;DR: Pods seem to be querying DNS servers in /etc/resolv.conf in random or round-robin order instead of sequentially, as they are supposed to.

Long version:
So the root of all of this is the DNAT race condition outlined in detail here: https://github.com/kubernetes/kubernetes/issues/56903
One of the recommended solutions is to run node-local DNS to prevent the race condition. However, if you go down that path its recommended to have a backup DNS in case the local DNS pod fails.  In my case I'm trying to run a node-local CoreDNS daemonset available on 192.168.0.10, and the cluster standard DNS server (in our case also CoreDNS) as a backup at 100.64.0.10.  I'm using kops for deployment, so I've simply added:
```
  kubelet:
    clusterDNS: 192.168.0.10,100.64.0.10
```
to the cluster spec.  That part appears to work since when I `kubectl exec` into a cluster and `cat /etc/resolv.conf` I get:
```
# cat /etc/resolv.conf
nameserver 192.168.0.10
nameserver 100.64.0.10
search kube-system.svc.cluster.local svc.cluster.local cluster.local us-west-2.compute.internal
options ndots:5
```
Everything looks good.  But I have a few simple CronJobs that run every minute and just perform various `dig` calls.  When I look at the logs from these jobs, it's clear that its still using **BOTH** node-local and cluster DNS to resolve names:
![image](https://user-images.githubusercontent.com/15616541/58057206-59120080-7b21-11e9-84d8-b343a9e8ec15.png)
Theres a picture of logs from an Alpine container running dig.  You can see the `SERVER` log line clearly is using both local DNS (192.168.0.10) and cluster DNS (100.64.0.10), and at a substantial rate - about 25% of the time.  It's similar with an Ubuntu-based image:
![image](https://user-images.githubusercontent.com/15616541/58057321-ae4e1200-7b21-11e9-894e-ea1fc27d6e21.png)
Ubuntu is all the way up at 35%. I would expect local DNS (192.168.0.10) to be 99%+ and cluster DNS (100.64.0.10) to be <1% since local DNS should **always** be queried first and only cluster DNS if local is unreachable.

The man page for /etc/resolv.conf (http://man7.org/linux/man-pages/man5/resolv.conf.5.html) clearly states that:
`If there are multiple servers, the resolver library queries them in the order listed`
But that is not happening here.  Unfortunately it seems I'm stuck... The only way to get rid of all the DNS timeouts is to completely remove cluster DNS from the kubelet config. So it appears I have to choose between a highly available DNS solution (with timeouts) or a local DNS solution (without high availability).

What's concerning is that it appears that the containers aren't respecting the intended design of /etc/resolv.conf with respect to nameserver ordering. And unfortunately there are so many layers here that I don't know if its a library issue (glibc vs musl), Linux kernel issue, Kubernetes issue or kops issue. If anybody has any insight on this it would be appreciated.

**What you expected to happen**:
Pods to query DNS servers in the order listed in /etc/resolv.conf

**How to reproduce it (as minimally and precisely as possible)**:
Configure cluster with multiple DNS servers in kubelet and perform `dig` lookups.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.0", GitCommit:"91e7b4fd31fcd3d5f436da26c980becec37ceefe", GitTreeState:"clean", BuildDate:"2018-06-27T20:17:28Z", GoVersion:"go1.10.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.9", GitCommit:"16236ce91790d4c75b79f6ce96841db1c843e7d2", GitTreeState:"clean", BuildDate:"2019-03-25T06:30:48Z", GoVersion:"go1.10.8", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:
AWS
- OS (e.g: `cat /etc/os-release`):
Using Kops Debian-stretch image:
```
PRETTY_NAME="Debian GNU/Linux 9 (stretch)"
NAME="Debian GNU/Linux"
VERSION_ID="9"
VERSION="9 (stretch)"
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```
- Kernel (e.g. `uname -a`):
`Linux ip-172-20-52-21 4.9.0-9-amd64 #1 SMP Debian 4.9.168-1 (2019-04-12) x86_64 GNU/Linux`
- Install tools: 
Kops v. 1.11.1
- Network plugin and version (if this is a network-related bug): 
weave (weaveworks/weave-npc:2.5.0)
- Others:

@kubernetes/sig-network-bugs

**Comment by chrisghill** [comment]
More information: I modified one of the DNS CronJobs to `echo \`cat /etc/resolv.conf\`` so that I could see the raw output on one line in order to ensure nameserver order.  I also changed the command to `dig +trace www.github.com`.  Here is the output:

```
nameserver 192.168.0.10 nameserver 100.64.0.10 search prod-ops.svc.cluster.local svc.cluster.local cluster.local us-west-2.compute.internal options ndots:5

; <<>> DiG 9.9.5-3ubuntu0.2-Ubuntu <<>> +trace www.github.com
;; global options: +cmd
.                  30      IN      NS      i.root-servers.net.
.                  30      IN      NS      j.root-servers.net.
.                  30      IN      NS      k.root-servers.net.
.                  30      IN      NS      l.root-servers.net.
.                  30      IN      NS      m.root-servers.net.
.                  30      IN      NS      a.root-servers.net.
.                  30      IN      NS      b.root-servers.net.
.                  30      IN      NS      c.root-servers.net.
.                  30      IN      NS      d.root-servers.net.
.                  30      IN      NS      e.root-servers.net.
.                  30      IN      NS      f.root-servers.net.
.                  30      IN      NS      g.root-servers.net.
.                  30      IN      NS      h.root-servers.net.
;; Received 420 bytes from 100.64.0.10#53(100.64.0.10) in 31 ms

com.                       172800  IN      NS      e.gtld-servers.net.
com.                       172800  IN      NS      b.gtld-servers.net.
com.                       172800  IN      NS      j.gtld-servers.net.
com.                       172800  IN      NS      m.gtld-servers.net.
com.                       172800  IN      NS      i.gtld-servers.net.
com.                       172800  IN      NS      f.gtld-servers.net.
com.                       172800  IN      NS      a.gtld-servers.net.
com.                       172800  IN      NS      g.gtld-servers.net.
com.                       172800  IN      NS      h.gtld-servers.net.
com.                       172800  IN      NS      l.gtld-servers.net.
com.                       172800  IN      NS      k.gtld-servers.net.
com.                       172800  IN      NS      c.gtld-servers.net.
com.                       172800  IN      NS      d.gtld-servers.net.
com.                       86400   IN      DS      30909 8 2 E2D3C916F6DEEAC73294E8268FB5885044A833FC5459588F4A9184CF C41A5766
com.                       86400   IN      RRSIG   DS 8 1 86400 20190602210000 20190520200000 25266 . KwfpxNtPTdyTfeISGuTw+BIbDiOYUw0wkvBoVFlyJpiRghKiNeairm/f qR688rzZ+c/c9LeK5pKDAg0NfbM0D6hGdfwKL0TpY5O6j0r9bLvBb07e 4yhg9Fl8q1DdtZj30q7b4RwqzuJWX5YrKd0Ve4z45JEJNlWjCG/UqImC 1WkzoXX/7q4287d9GixyVk9OR58BfX6lSOAWM9mR9pdJeBGmJF5RLNlK LSDMCQ2ABYd3xlzBw2LpxAdjO6FOHhHsEoWZzZh5MKq2TmPzsX4/FBNR 36ZrjUHFI7g3cEsp2WPNW7otJP3vOgRnLwP0ZiF5+iOsL13LXscWrFg9 V0Ysfg==
;; Received 1174 bytes from 198.41.0.4#53(a.root-servers.net) in 50 ms

github.com.                172800  IN      NS      ns1.p16.dynect.net.
github.com.                172800  IN      NS      ns3.p16.dynect.net.
github.com.                172800  IN      NS      ns2.p16.dynect.net.
github.com.                172800  IN      NS      ns4.p16.dynect.net.
github.com.                172800  IN      NS      ns-520.awsdns-01.net.
github.com.                172800  IN      NS      ns-421.awsdns-52.com.
github.com.                172800  IN      NS      ns-1707.awsdns-21.co.uk.
github.com.                172800  IN      NS      ns-1283.awsdns-32.org.
CK0POJMG874LJREF7EFN8430QVIT8BSM.com. 86400 IN NSEC3 1 1 0 - CK0Q1GIN43N1ARRC9OSM6QPQR81H5M9A NS SOA RRSIG DNSKEY NSEC3PARAM
CK0POJMG874LJREF7EFN8430QVIT8BSM.com. 86400 IN RRSIG NSEC3 8 2 86400 20190527044553 20190520033553 3800 com. kcAfr7kjt62v1h38aMUqjbgnIPH5ayOSRGT26OvsLnoF5/EwerLgu1oO 8nMq4lbDrlJO0+zPPKVvv6ELOthJ4euYrFZkw2YjidlBRlxpQHm2Lgjh ztHDNsPUmjPrqavV0Xorq8WWy7gkabmfcrx9EJEN4xgv52GTeD76TnKy geQ=
4KB49MHL0SB1JJLOK123N91V2C905VHU.com. 86400 IN NSEC3 1 1 0 - 4KB4PTQQ5CTA7POCTGM7RUFC8B1RKTEU NS DS RRSIG
4KB49MHL0SB1JJLOK123N91V2C905VHU.com. 86400 IN RRSIG NSEC3 8 2 86400 20190526052051 20190519041051 3800 com. GETnHOwC9h9siBEFbQhM0OzVMMA/UMN/6zQtFKIvVaiWaBPoSxusfka1 6LYUTtTvkHth2L6lhAnCasj8M5218RAedSd8HYWSXOM5TywwePimgTvQ pj0rhYtUliaWPzZ6B/2/K4OGL+t2yojMItXXH/fIr9QYSYmVVpHFnPye n4E=
;; Received 764 bytes from 192.35.51.30#53(f.gtld-servers.net) in 132 ms

www.github.com.            3600    IN      CNAME   github.com.
github.com.                60      IN      A       192.30.255.113
github.com.                900     IN      NS      ns-421.awsdns-52.com.
github.com.                900     IN      NS      ns-520.awsdns-01.net.
github.com.                900     IN      NS      ns4.p16.dynect.net.
github.com.                900     IN      NS      ns-1283.awsdns-32.org.
github.com.                900     IN      NS      ns2.p16.dynect.net.
github.com.                900     IN      NS      ns-1707.awsdns-21.co.uk.
github.com.                900     IN      NS      ns1.p16.dynect.net.
github.com.                900     IN      NS      ns3.p16.dynect.net.
;; Received 293 bytes from 208.78.71.16#53(ns3.p16.dynect.net) in 8 ms
```

You can clearly see that server 192.168.0.10 is listed as the first DNS server, but the `dig +trace` goes straight to 100.64.0.10.

**Comment by athenabot** [comment]
/triage unresolved

Comment `/remove-triage unresolved` when the issue is assessed and confirmed.

🤖 I am a bot run by vllry. 👩‍🔬

**Comment by prameshj** [comment]
Setting multiple nameservers in /etc/resolv.conf can have inconsistent behavior based on the client library in the pod. I have not tried all combinations, but musl sends queries in parallel to both nameservers.  There are some HA options listed here: 
https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/20190424-NodeLocalDNS-beta-proposal.md#proposal

**Comment by prameshj** [comment]
/kind feature
feature request for nodelocaldns HA.

**Comment by alwaysused** [comment]
Actually  you got msg on terminal only when the dig get response from server.The process is that dig first referred to the 192... and  after  timeout it referred to the next server in resolve.conf,10...,then dig got response and printed it on the terminal.Maybe you should use some linux commands instead of  dig trace.

**Comment by prameshj** [comment]
@chrisghill Is there anything more needed here?
/assign @chrisghill

**Comment by chrisghill** [comment]
@prameshj No, I was able to solve the timeout through a different means, though I do see value in having HA nodelocaldns.  Thanks for following up.  I'll go ahead and close the issue.

**Comment by george-angel** [comment]
@chrisghill Would you mind sharing how you resolved this issue?

We have currently stopped the roll-out of nodelocaldns because of this issue. Thanks!

**Comment by johnbelamaric** [comment]
@george-angel If you are using a musl-based distro (e.g., Alpine), it will issue request to all nameservers in parallel. There's nothing we can do on the server side about that.

**Comment by chrisghill** [comment]
@george-angel possibly. I wasn't able to actually solve the DNS server lookup-order issue that this ticket was originally about.  From the comments here it sounds like that would be difficult since Alpine uses musl which issues requests to all nameservers at once.  But, the only reason I was exploring nodelocaldns was because it was a suggested fix to the DNS server timeout issue discussed at length here: https://github.com/kubernetes/kubernetes/issues/56903

If you're wondering how I solved the DNS timeout issue, theres a suggested fix by Quentin Machu  who just added a 3 ms delay to AAAA lookups to prevent the A/AAAA lookup race condition that was the source of the timeouts.  His fix is detailed and made available here: https://github.com/Quentin-M/weave-tc

**Comment by johnbelamaric** [comment]
You can blackhole the AAAA requests in CoreDNS with a rewrite rule, something like this will likely work: `rewrite type AAAA X25`. Basically that rewrites any AAAA request to a request for an X25 record, which is highly unlikely to actually exist (you could pick any other obscure record type). Thus the response is NODATA.

**Comment by chrisghill** [comment]
@johnbelamaric good to know.  However, from what I understand of the timeout issue, the source is in the Linux kernel libraries, and more importantly the race condition prevents the DNS requests from ever hitting a server at all.  At least that's what I pieced together from reading through the issue I linked.  I think that's why Quentin Machu's solution patches it at the network overlay level.  But if I'm wrong, then blackholing the AAAA requests at the DNS server may be another viable alternative.

**Comment by prameshj** [comment]
Thanks @chrisghill 
From your earlier comment in this issue I see:

"One of the recommended solutions is to run node-local DNS to prevent the race condition. However, if you go down that path its recommended to have a backup DNS in case the local DNS pod fails."
The recommendation to have a backup DNS - is that something you found in the docs somewhere or ran into an issue that required you to have a fallback? trying to understand the motivation for needing HA.

@george-angel Can you elaborate on the reason for stopping rollout of nodelocaldns? Is it because you need a fallback?

**Comment by chrisghill** [comment]
@prameshj I don't think I saw HA anywhere explicitly recommended, but in the DNS timeout issue I linked when other users were discussing their solutions and the topic of a DNS server per node came up the problem of HA was usually mentioned as well. I think it was just a recognition that at some point a local DNS server is likely to crash, causing failures in every pod on that node until it recovers. Not everyone was using the specific nodelocaldns addon - some tweaked CoreDNS to be a daemonset, some used a dnsmasq daemonset with kube-dns as the upstream server. I borrowed somebody else's solution which was to run a CoreDNS daemonset and update kubelet to check the local DNS first and cluster CoreDNS 2nd.  But obviously there were issues with that, hence this ticket.  I ended up abandoning that approach and just solved the timeouts by adding the delay to AAAA lookups.  I'm back to just running a standard CoreDNS deployment with 2+ replicas.

**Comment by george-angel** [comment]
@prameshj correct, because using node-local DNS alone means every time there is a deployment (adjust resources, new version, etc) - there is downtime for dns resolution on the node. So in my mind having a fallback is a must.

**Comment by prameshj** [comment]
Thanks @george-angel , @chrisghill . I had submitted a KEP that talks about an HA option using the kube-dns service ip as the listen ip for the cache. 
https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/20190424-NodeLocalDNS-beta-proposal.md#design-details

The solution is not fully available in the latest node-cache version, but I did introduce a few commandline flags which will enable setting up what is described in the KEP. It would be great to get your feedback on this approach.
